### DISCUSSION TITLE
[Added] XLink Kai

### DISCUSSION THREAD

**Discussion by sairuk** [retronas/discussions#39] [General]
XLink Kai is now available in a branch, it needs testing. I still haven't received the auth response to my registration with the project so I can only so some limited testing until that comes through. if you already have an account and can test it out that would be helpful

Use the branch switcher from the Updates menu to switch to the xlinkkai branch, the installer is available in the Experimental menu.

[Documentation](https://github.com/danmons/retronas/wiki/XLink-Kai)

**Comment by s3anami** [comment]
I can try and test but does it have any known  conflicts with other functions and can I switch branches back without issue?

**Reply by sairuk** [reply]
no known conflicts and yes you should be able to switch back to main without issue using the same tool from the update menu

xlink will however remain installed once you switch back to main.

i have my account now so will be able to test at some point. I might try with xemu

**Reply by s3anami** [reply]
i tried with xemu and my og xbox couldn't get it to work
seems to crash when doing lots of tasks too and some configuration settings

**Reply by sairuk** [reply]
can you provide details on what to do to replicate the crash so we can see if it is a retronas or xlink thing

**Reply by s3anami** [reply]
I was just trying different settings to see if I could make it work, devices are seen as configured ok under the metrics section.
If I logout of my account the web console becomes inaccessible, if I try to save when changing certain configurations (I was messing around trying to get it to work), for example Allow Remote DDS Connections when I save the settings it crashes. 
Settings in the configuration are also not retained on restart of system.

**Reply by sairuk** [reply]
default perms on the config from the installer are not correct, this is part of the crashing issue

**Reply by s3anami** [reply]
> default perms on the config from the installer are not correct, this is part of the crashing issue

ok let me know what else I can do to help with testing and when things are ready to test, seems like xlink can see my systems and everything appears normal with people in the lobbies connection health being seen but can't see anything in system link. Was testing using halo 2

**Reply by sairuk** [reply]
i am seeing similar behaviour playing around with rainbow six 3, everything looks ok but it doesn't comm properly. The help dialog for console setup seems relatively generic but I'll spend some more time with it.

it seems kai only uses the config in `/etc` by default so `chmod 666 /etc/kaiengine.conf` lets the config be updated from the webui as the retronas user

a lot of this seems like its going to be outside of RetroNAS control, we run the application service and the rest is dependent on individual network setup and application stability my goal here is understand and document what it'll take to get it working.

**Reply by sairuk** [reply]
so i have this working now, to play there's nothing extra to do it just works

1. Login to XLink
2. Browse to a console, game and region with players
3. Start Halo 2
4. Select Profile
5. Select System Link
6. Join
7. Browse/Create/Join

Running servers will show up and you can join/play without issue, no port forwards were required to play.

Tested on both an RPI and a VM, both worked OK

I will be doing fixed ports for hosting because I don't allow UPnP and I am double NAT'd here. I know my game showed up for someone but they couldn't join the lobby because my transfer was bad of my Halo 2 game had an error loading maps. I've fixed that now. So I need to retest

**Reply by s3anami** [reply]
ok i will reinstall the xlink updated portion and retry

**Reply by s3anami** [reply]
> 

i can retest with you if you want, but I haven't been able to see any games not sure if its me or the lobbies made are dead, i did the same as you with static port for port forwarding

**Reply by sairuk** [reply]
i've pushed some changes today

* the xlinkkai options are now in testing branch instead of their own.
* there is now an xlinkkai menu where you can manage the service
* xlinkkai service will now restart always which should help with the odd segfaults
* xlinkkai service will stop cleanly now sending a SIGQUIT to the main process

the service now restarts for me consistently and I was able to play some Halo 2 last night (badly)

**Reply by s3anami** [reply]
what hyper visor do you have? and are you using promiscuous networking mode?

**Reply by sairuk** [reply]
esxi and no to prom. mode. also no particular setup here I can think of. 

make sure you're on when there are games running in a region/arena for the right title. you won't see games listed in the web ui, will only see them ingame

**Reply by s3anami** [reply]
yup. I have used the windows version before, so I am familiar with the way it works. It seems that i needed to have promiscuous mode turned on for the vm to broadcast appropriately to my original xbox and an emulator not on the host. I can see games now but haven't been able to test joining one since they are mostly in progress

**Reply by sairuk** [reply]
interesting, good to know it is now working for you. 

I took a bit more of a look this morning, while my esxi net doesn't have prom mode on, nor do my ints in the vm the devices do switch state to prom mode

### rpi

ip a
```
2: eth0: <BROADCAST,MULTICAST,ALLMULTI,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
3: wlan0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
```
dmesg
```
[15031.352319] device eth0 entered promiscuous mode
[15031.376536] device wlan0 entered promiscuous mode
```

### vm

ip a
```
2: ens192: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
```

dmesg
```
[4726713.230718] device ens192 left promiscuous mode
[4726723.338220] device ens192 entered promiscuous mode
```

![image](https://user-images.githubusercontent.com/526372/180885350-7602de18-5743-4f85-bdad-e8f3eeacb90d.png)
it is vmware tho so i'm not surprised it doesn't honor settings

**Reply by s3anami** [reply]
that is concerning lol
Anyway for Hyper-V  it mostly needs to be done via CLI too (ironic, it being windows based and esxi being linux)
I did the first 8 steps here:
https://docs.tenable.com/nnm/deployment/Content/VM/Hyper-VInternal.htm

**Reply by sairuk** [reply]
so a little bit more reading and it claims the vswitch/portgroup policy is working but as xlink is working i have my doubts. i'd need to start looking at traffic when I have time but since i have no expectations of running this on a vm motivation is low, I was already planning on dropping esxi.

```
2022-07-25T20:37:52.046Z cpu5:2100739)cswitch: L2Sec_EnforcePortCompliance:207: [nsx@6876 comp="nsx-esx" subcomp="vswitch"]client  requested promiscuous mode on port 0x200000e, disallowed by vswitch policy
```

From a RetroNAS perspective we just need to document prom support is required for functionality and move on.

**Reply by s3anami** [reply]
sounds fine, ill let you know if there is anything else i run into when trying it, only got into 1 latam halo game so far and was swiftly kicked so  ill keep trying

**Reply by sairuk** [reply]
thanks for your help with testing this

**Reply by teamxlink-dev** [reply]
You'll only _need_ forged transmits for Xbox and Xbox 360 games. For the other platforms we have an internal workaround where it is not required. Promiscuous mode may not be needed for all platforms but Kai will still try to enable it

**Comment by sairuk** [comment]
When logging out of the webui running as a normal user

```
kaiengine: /opt/xlink/kaiEngine/references_aarch64/boost_1_76_0/boost/thread/pthread/recursive_mutex.hpp:108: void boost::recursive_mutex::lock(): Assertion `!posix::pthread_mutex_lock(&m)' failed.
Interrupt signal (6) received.
Segmentation fault
```

**Reply by sairuk** [reply]
but only if kaiUIClose=1 in the conf otherwise logout will exit the engine cleanly (exit-code 0) which is not the expected behaviour.

So logout from the web-ui is actually an exit action by default, we can probably work around this by using `Restart=always`
 in the systemd service file and then providing an exception for select exit codes

**Comment by sairuk** [comment]
Played around with Xemu in Bridged mode for this as well, 

* Ubuntu install, worked fine with the `setcap` perms fix from the official doco
* Manjaro `aur` or `flatpak` both reporting `"Address family not supported by protocol"` which I haven't resolved yet

**Comment by noz1380** [comment]
Just had a play on Halo 2, seemed easy to setup and find a game.

I'll have a play with my PSP and Vita too.

Thank you for all the hard work, this is awesome.

**Comment by teamxlink-dev** [comment]
Hey there, if you have any questions let us know :)

I've read through your XLink wiki entry and you've hit a few known bugs (erroneous IP incorrect etc)

**Reply by sairuk** [reply]
Will do, thanks for taking a look at our doco
